### PR TITLE
Adding test for compute node pod whitelist.

### PR DESCRIPTION
**Pull request by kwoodson**
Adding a test for pods on compute nodes.  The basic gist is that we are verifying that infra node pods do not get deployed onto compute nodes.  We are creating a whitelist of pods allowed on the compute nodes.

fixes: https://github.com/Azure/acs-engine/issues/2895

@jim-minter @kargakis

### CODE REVIEW

**Comment by kargakis** [comment]
+1 for blacklisting

**Comment by kwoodson** [comment]
@kargakis, 

@jim-minter and I had a chat about a strategy here.  We wanted to whitelist namespaces as well as pods. 

I've also added a test to check all deployed pods are not in an error state.

Please review.

**Comment by kargakis** [comment]
Sorry got this backwards

+1 for a whitelist

File: test/e2e/kubernetes/pod/pod.go [hunk 1]
```diff
@@ -107,9 +115,17 @@ func CreatePodFromFile(filename, name, namespace string) (*Pod, error) {
 	return pod, nil
 }
 
-// GetAll will return all pods in a given namespace
+// GetAll will return all pods in a given namespace. If namespace is "" will return all pods
 func GetAll(namespace string) (*List, error) {
-	cmd := exec.Command("kubectl", "get", "pods", "-n", namespace, "-o", "json")
+	command := []string{"get", "pods", "-o", "json"}
```
Comment (jim-minter) [inline test/e2e/kubernetes/pod/pod.go]: s/command/args/

File: test/e2e/openshift/openshift_test.go [hunk 2]
```diff
@@ -85,6 +85,62 @@ var _ = Describe("Azure Container Cluster using the OpenShift Orchestrator", fun
...
+		}
+		list, err := pod.GetAll("")
+		Expect(err).NotTo(HaveOccurred())
+
+		for _, pod := range list.Pods {
+			skip := false
+			// we have a whitelist of namespaces that we will skip
+			for _, pattern := range namespaces {
+				if strings.Contains(pod.Metadata.Namespace, pattern) {
+					skip = true
```
Comment (jim-minter) [inline test/e2e/openshift/openshift_test.go]: `skip` is ugly - either using Go's `break LABEL` syntax (https://golang.org/ref/spec#Break_statements) or a separate function would be better.

File: test/e2e/openshift/openshift_test.go [hunk 3]
```diff
@@ -85,6 +85,62 @@ var _ = Describe("Azure Container Cluster using the OpenShift Orchestrator", fun
 		}
 	})
 
+	It("should whitelist pods on compute nodes", func() {
```
Comment (jim-minter) [inline test/e2e/openshift/openshift_test.go]: something like "should deploy all pods onto appropriate nodes" would be clearer?

File: test/e2e/openshift/openshift_test.go [hunk 4]
```diff
@@ -85,6 +85,62 @@ var _ = Describe("Azure Container Cluster using the OpenShift Orchestrator", fun
...
+		namespaces := []string{
+			"openshift-infra",
+			"openshift-node",
+			"openshift-metrics",
+			"kube-",
+			"default",
+		}
+
+		whitelist := []string{
```
Comment (jim-minter) [inline test/e2e/openshift/openshift_test.go]: daemonsetPrefixes

File: test/e2e/openshift/openshift_test.go [hunk 5]
```diff
@@ -85,6 +85,62 @@ var _ = Describe("Azure Container Cluster using the OpenShift Orchestrator", fun
...
+					skip = true
+					break
+				}
+			}
+			if skip {
+				skip = false
+				continue
+			}
+
+			found := false
```
Comment (jim-minter) [inline test/e2e/openshift/openshift_test.go]: move down a line

File: test/e2e/openshift/openshift_test.go (hunk 2 above)
Comment (jim-minter) [reply test/e2e/openshift/openshift_test.go]: why are we excluding certain namespaces?

File: test/e2e/openshift/openshift_test.go [hunk 6]
```diff
@@ -85,6 +85,62 @@ var _ = Describe("Azure Container Cluster using the OpenShift Orchestrator", fun
...
+	})
+
+	// Verify all pods have initialized, ready, podscheduled are true
+	It("should have all pods in ready state", func() {
+		list, err := pod.GetAll("")
+		Expect(err).NotTo(HaveOccurred())
+
+		for _, pod := range list.Pods {
+			for _, cond := range pod.Status.Conditions {
+				Expect(cond.Status).To(BeTrue())
```
Comment (jim-minter) [inline test/e2e/openshift/openshift_test.go]: too broad - just assert that Ready is true - leave the others

Comment (jim-minter) [reply test/e2e/openshift/openshift_test.go]: it might be easier to check .status.phase is Running or Completed than iterate the conditions

File: test/e2e/openshift/openshift_test.go [hunk 7]
```diff
@@ -85,6 +85,62 @@ var _ = Describe("Azure Container Cluster using the OpenShift Orchestrator", fun
...
+						break
+					}
+				}
+				Expect(found).To(BeTrue())
+			}
+		}
+	})
+
+	// Verify all pods have initialized, ready, podscheduled are true
+	It("should have all pods in ready state", func() {
```
Comment (jim-minter) [inline test/e2e/openshift/openshift_test.go]: have you tested this passes against a cluster today? does it catch the asb problem that we've currently got? if so, to get this in, you might need to exclude the asb pods from the test and add a `// TODO: remove this when issue #xyz is closed`

File: test/e2e/openshift/openshift_test.go [hunk 8]
```diff
@@ -85,6 +85,62 @@ var _ = Describe("Azure Container Cluster using the OpenShift Orchestrator", fun
...
+				}
+			}
+			if skip {
+				skip = false
+				continue
+			}
+
+			found := false
+			if strings.Contains(pod.Spec.Nodename, "compute") {
+				for _, pattern := range whitelist {
```
Comment (jim-minter) [inline test/e2e/openshift/openshift_test.go]: think about ensuring that the intent of your test is clear to a 3rd party.  What you're testing is: "no pods should be on a compute node unless they start with a daemonsetPrefix".  Try to use patterns and variable names that make this clear, e.g. perhaps something like "isDaemonsetPod := matchPrefix()... ; assert(isDaemonsetPod)"

File: test/e2e/openshift/openshift_test.go [hunk 9]
```diff
@@ -85,6 +85,62 @@ var _ = Describe("Azure Container Cluster using the OpenShift Orchestrator", fun
...
+					if strings.Contains(pod.Metadata.Name, pattern) {
```
Comment (jim-minter) [inline test/e2e/openshift/openshift_test.go]: strings.HasPrefix

File: test/e2e/openshift/openshift_test.go (hunk 3 above)
Comment (kwoodson) [reply test/e2e/openshift/openshift_test.go]: Currently we are only checking `compute` node types.  Should we add master and infra types as well?

File: test/e2e/kubernetes/pod/pod.go (hunk 1 above)
Comment (kwoodson) [reply test/e2e/kubernetes/pod/pod.go]: Updated.

File: test/e2e/openshift/openshift_test.go (hunk 2 above)
Comment (kwoodson) [reply test/e2e/openshift/openshift_test.go]: The idea was that any pod that is in the openshift-metrics namespace we want to skip regardless of host.  I see that this might bypass too large a space with the list that I came up with.  I will only skip `openshift-metrics` currently.

File: test/e2e/openshift/openshift_test.go (hunk 4 above)
Comment (kwoodson) [reply test/e2e/openshift/openshift_test.go]: This could also include `statefulsets`.  Do you still prefer `daemonsetPrefixes`?

File: test/e2e/openshift/openshift_test.go (hunk 9 above)
Comment (kwoodson) [reply test/e2e/openshift/openshift_test.go]: Updated.

File: test/e2e/openshift/openshift_test.go (hunk 8 above)
Comment (kwoodson) [reply test/e2e/openshift/openshift_test.go]: This makes sense.  Will update.

File: test/e2e/openshift/openshift_test.go (hunk 5 above)
Comment (kwoodson) [reply test/e2e/openshift/openshift_test.go]: done.

File: test/e2e/openshift/openshift_test.go (hunk 6 above)
Comment (kwoodson) [reply test/e2e/openshift/openshift_test.go]: There are 3 condition types:
1.  Initialized
2.  Ready
3.  PodScheduled

All _should_ be `true`.  Would you prefer if we only check for `Ready`?

**Comment by kargakis** [comment]
There is a PodReady helper we use elsewhere in tests. Checking for pod
readiness covers initialization and scheduling.

File: test/e2e/openshift/openshift_test.go (hunk 6 above)
Comment (kwoodson) [reply test/e2e/openshift/openshift_test.go]: I checked on a cluster with an application in a CrashLoopBackOff  state:
```
  conditions:
  - lastProbeTime: null
    lastTransitionTime: 2018-06-05T02:41:43Z
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: 2018-06-12T18:34:06Z
    message: 'containers with unready status: [glass]'
    reason: ContainersNotReady
    status: "False"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: 2018-06-05T02:41:43Z
    status: "True"
    type: PodScheduled
```

The `phase` was set to `Running`.
```
  phase: Running
```

I believe that phase wouldn't be sufficient.

File: test/e2e/openshift/openshift_test.go (hunk 7 above)
Comment (kwoodson) [reply test/e2e/openshift/openshift_test.go]: Will run and confirm that we catch this case.

**Comment by kargakis** [comment]
Neither 'ready = true' is sufficient when the pod is crashlooping, that's
why the helper used in the tests checks for readiness 3-4 times in a row.

**Comment by kwoodson** [comment]
@kargakis, does the PodReady helper find the asb pod that is stuck waiting for the image to deploy?  If it does then we can remove those tests.

**Comment by kargakis** [comment]
We already have readiness tests for all infra pods and yes the asb test
gets stuck in 3.10. We should either have your test check only that pods
got scheduled on the correct nodes or collapse with the existing tests if
you also want to check readiness.

**Comment by kwoodson** [comment]
As per @kargakis' advice, I have removed the ready test as those are implemented elsewhere.

I also ended up moving the nginx test to `e2e-test` so that the whitelist testing works.
```
Azure Container Cluster using the OpenShift Orchestrator 
  should deploy all pods onto appropriate nodes
  /go/src/github.com/Azure/acs-engine/test/e2e/openshift/openshift_test.go:89

$ kubectl get pods -o json --all-namespaces
•
------------------------------
```
The nginx test worked as well:
```
Azure Container Cluster using the OpenShift Orchestrator 
  should deploy a sample app and access it via a route
  /go/src/github.com/Azure/acs-engine/test/e2e/openshift/openshift_test.go:175

$ kubectl create namespace e2e-test

$ kubectl get namespace e2e-test -o json

$ oc process nginx-example -n openshift

$ oc apply -n e2e-test -f nginx-example

$ oc rollout status dc/nginx-example -n e2e-test

$ oc get route/nginx-example -n e2e-test -o jsonpath={.spec.host}
•
------------------------------
```

The namespace and pods:
```
[root@ocp-master-94252885-0 ~]# oc get pods -n e2e-test
NAME                    READY     STATUS      RESTARTS   AGE
nginx-example-1-build   0/1       Completed   0          21m
nginx-example-1-wbc98   1/1       Running     0          20m
```

File: test/e2e/openshift/openshift_test.go [hunk 10]
```diff
@@ -85,6 +86,44 @@ var _ = Describe("Azure Container Cluster using the OpenShift Orchestrator", fun
 		}
 	})
 
+	It("should deploy all pods onto appropriate nodes", func() {
+		namespaces := []string{
+			"openshift-metrics",
```
Comment (jim-minter) [inline test/e2e/openshift/openshift_test.go]: I think you can add prometheus-node-exporter- to daemonsetPrefixes and get rid of openshift-metrics here?
The test should then say:
1) if the pod is from a daemonset, don't care about its node.
2) if the pod's name starts registry-console-, don't care.
3) the pod must be on compute if and only if the pod's namespace does not start with openshift-, kube- and the namespace is not "default"

File: test/e2e/openshift/openshift_test.go [hunk 11]
```diff
@@ -85,6 +86,44 @@ var _ = Describe("Azure Container Cluster using the OpenShift Orchestrator", fun
...
+			"e2e-test",
+		}
+
+		daemonsetPrefixes := []string{
+			"sync-",
+			"ovs-",
+			"sdn-",
+			"registry-console", // Remove this line once registry-console adds the proper nodeSelector
```
Comment (jim-minter) [inline test/e2e/openshift/openshift_test.go]: reference issue

File: test/e2e/openshift/openshift_test.go [hunk 12]
```diff
@@ -85,6 +86,44 @@ var _ = Describe("Azure Container Cluster using the OpenShift Orchestrator", fun
 		}
 	})
 
+	It("should deploy all pods onto appropriate nodes", func() {
+		namespaces := []string{
+			"openshift-metrics",
+			"e2e-test",
+		}
+
+		daemonsetPrefixes := []string{
```
Comment (jim-minter) [inline test/e2e/openshift/openshift_test.go]: ideally you would get rid of daemonsetPrefixes entirely; use "pod has an OwnerReference of kind DaemonSet" to detect a pod owned by a daemonset.

**Comment by kwoodson** [comment]
@jim-minter, implemented your latest strategy

> 1 if the pod is from a daemonset, don't care about its node.
> 2 if the pod's name starts registry-console-, don't care.
> 3 the pod must be on compute if and only if the pod's namespace does not start with openshift-, kube- > and the namespace is not "default"

```
Azure Container Cluster using the OpenShift Orchestrator 
  should deploy all pods onto appropriate nodes
  /go/src/github.com/Azure/acs-engine/test/e2e/openshift/openshift_test.go:89

$ kubectl get pods -o json --all-namespaces
•
------------------------------
```

```
[root@ocp-master-18458924-0 ~]# oc adm manage-node ocp-compute-18458924-0  --list-pods

Listing matched pods on node: ocp-compute-18458924-0

NAMESPACE           NAME                             READY     STATUS    RESTARTS   AGE
openshift-metrics   prometheus-node-exporter-l6nmj   1/1       Running   0          21m
openshift-node      sync-9m845                       1/1       Running   0          22m
openshift-sdn       ovs-wmss5                        1/1       Running   0          22m
openshift-sdn       sdn-xkxq7                        1/1       Running   0          22m
```

**Comment by kwoodson** [comment]
@jim-minter, please review.

File: test/e2e/openshift/openshift_test.go [hunk 13]
```diff
@@ -85,6 +86,38 @@ var _ = Describe("Azure Container Cluster using the OpenShift Orchestrator", fun
 		}
 	})
 
+	It("should deploy all pods onto appropriate nodes", func() {
```
Comment (kargakis) [inline test/e2e/openshift/openshift_test.go]: Move your test at the end of the suite to ensure the pod list is not racing with the deployer pods and you are getting all pods we expect to run.

Comment (kargakis) [reply test/e2e/openshift/openshift_test.go]: Also I would reword the title of the test to "should not deploy infrastructure components in compute nodes" since you check just that (eg. you don't discriminate between pods that land in master vs infra).

Comment (kwoodson) [reply test/e2e/openshift/openshift_test.go]: Ok, I will move this to the end of the tests.  Will also rename.

**Comment by kwoodson** [comment]
@kargakis, updated.

**Comment by kargakis** [comment]
/lgtm

@CecileRobertMichon please approve

**Comment by CecileRobertMichon** [comment]
@kwoodson please fix the lint error in the unit tests

**Comment by kwoodson** [comment]
@kargakis, @CecileRobertMichon, Added doc string to clean up linting.

**Comment by kwoodson** [comment]
Rebasing onto master.

**Comment by kargakis** [comment]
@CecileRobertMichon please approve

**Comment by kwoodson** [comment]
@kargakis, @jim-minter,

Have you seen this on our e2e?
```
Error from deployment for openshift-eastus-80970 in resource group openshift-eastus-80970:exit status 1
2018/06/21 20:52:16 Command Output: ERROR: The template deployment 'openshift-eastus-80970' is not valid according to the validation procedure. The tracking id is 'd5713f39-0adb-4318-93be-ded82aab76fe'. See inner errors for details. Please see https://aka.ms/arm-deploy for usage details.
```

**Comment by kargakis** [comment]
Nothing in the activity log .......... Unclear whether we would get more info out of the returned tracking ID if we can't get any activity log from the whole resource group.

File: test/e2e/openshift/openshift_test.go [hunk 14]
```diff
@@ -171,4 +175,36 @@ var _ = Describe("Azure Container Cluster using the OpenShift Orchestrator", fun
...
+		namespaces := []string{
+			"openshift-",
+			"kube-",
+			"default",
+		}
+
+		list, err := pod.GetAll("")
+		Expect(err).NotTo(HaveOccurred())
+
+		for _, pod := range list.Pods {
```
Comment (jim-minter) [inline test/e2e/openshift/openshift_test.go]:
```
nextPod:
		for _, pod := range list.Pods {
			// skip if we find a pod that is a daemonset
			for _, ref := range pod.Metadata.OwnerReferences {
				if ref.Kind == "DaemonSet" {
					continue nextPod
				}
			}
```

^ is better because (a) it would handle multiple ownerreferences if they ever happened for some reason, and (b) testing for a non-nil slice does not imply that the slice has an element (nb you can use len > 0 for that since len(nil slice) is defined to be 0)

File: test/e2e/openshift/openshift_test.go [hunk 15]
```diff
@@ -171,4 +175,36 @@ var _ = Describe("Azure Container Cluster using the OpenShift Orchestrator", fun
...
+			// Needs to be configurable.  Skip for now.
+			// https://github.com/Azure/acs-engine/issues/2895
+			if strings.HasPrefix(pod.Metadata.Name, "registry-console") {
+				continue
+			}
+
+			if strings.Contains(pod.Spec.Nodename, "compute") {
+				for _, pattern := range namespaces {
+					if strings.HasPrefix(pod.Metadata.Namespace, pattern) {
+						Expect(false).To(BeTrue())
```
Comment (jim-minter) [inline test/e2e/openshift/openshift_test.go]: `Expect(strings.HasPrefix(pod.Metadata.Namespace, pattern)).To(BeFalse())`

**Comment by jim-minter** [comment]
@kwoodson 

> Have you seen this on our e2e?

Open a flake issue with link to the job if it recurs.

**Comment by kargakis** [comment]
@kwoodson reopen and rebase if you are planning to wrap this up. Not sure we need it anymore.

/close
